include/dr/mhp/containers/rows.hpp

```diff
@@ -0,0 +1,172 @@
+
+  dm_rows_iterator() = default;
+  dm_rows_iterator(DM *dm, std::size_t index) noexcept {
+    dm_ = dm;
+    row_idx_ = index;
+  }
+
+  auto operator*() const { return dm_->dm_rows_[row_idx_]; }
+  auto operator->() const { return &(dm_->dm_rows_[row_idx_]); }
+  auto operator[](difference_type n) const {
+    difference_type abs_ind = row_idx_ + n;
+
+    if (abs_ind >= dm_->local_rows_ind_.first &&
+        abs_ind <= dm_->local_rows_ind_.second) { // regular rows
+      return dm_->dm_rows_[row_idx_ + n];
+    }
+    if (abs_ind >= (difference_type)(dm_->local_rows_ind_.first -
+                                     dm_->halo_bounds_rows_.prev) &&
+        abs_ind < dm_->local_rows_ind_.first) { // halo prev
+      return dm_->dm_halo_p_rows_[dm_->halo_bounds_rows_.prev -
+                                  dm_->local_rows_ind_.first + abs_ind];
+    }
+    if (abs_ind > dm_->local_rows_ind_.second &&
+        abs_ind <=
+            (difference_type)(dm_->local_rows_ind_.second +
+                              dm_->halo_bounds_rows_.next)) { // halo next
+      return dm_->dm_halo_n_rows_[dm_->halo_bounds_rows_.next +
+                                  dm_->local_rows_ind_.second - abs_ind];
+    }
+    assert(0);
+  }
+
+  // Comparison
+  bool operator==(const dm_rows_iterator &other) const noexcept {
+    return row_idx_ == other.row_idx_ && dm_ == other.dm_;
+  }
+  auto operator<=>(const dm_rows_iterator &other) const noexcept {
+    return row_idx_ <=> other.row_idx_;
+  }
+
+  // Only these arithmetics manipulate internal state
+  auto &operator-=(difference_type n) {
+    row_idx_ -= n;
+    return *this;
+  }
+  auto &operator+=(difference_type n) {
+    row_idx_ += n;
+    return *this;
+  }
+  difference_type operator-(const dm_rows_iterator &other) const noexcept {
+    return row_idx_ - other.row_idx_;
+  }
+
+  // prefix
+  auto &operator++() {
+    *this += 1;
+    return *this;
+  }
+  auto &operator--() {
+    *this -= 1;
+    return *this;
+  }
+
+  // postfix
+  auto operator++(int) {
+    auto prev = *this;
+    *this += 1;
+    return prev;
+  }
+  auto operator--(int) {
+    auto prev = *this;
+    *this -= 1;
+    return prev;
+  }
+
+  auto operator+(difference_type n) const {
+    auto p = *this;
+    p += n;
+    return p;
+  }
+  auto operator-(difference_type n) const {
+    auto p = *this;
+    p -= n;
+    return p;
+  }
+
+  // When *this is not first in the expression
+  friend auto operator+(difference_type n, const dm_rows_iterator &other) {
+    return other + n;
+  }
+
+  auto segments() { return dm_->segments(); }
+  auto rank() const { return (*this).rank(); }
+  auto &halo() const { return dm_->halo(); }
+
+  bool is_local() { return dm_->is_local_row(row_idx_); }
+
+  DM *dm() { return dm_; }
+
+private:
+  DM *dm_ = nullptr;
+  std::size_t row_idx_ = 0;
+}; // dm_rows_iterator
+
+template <typename DM>
+class dm_rows : public std::vector<dm_row<typename DM::value_type>> {
```

Comment:
Why do you inherit from std::vector? You don't initialize it.
I think you want this to be range adaptor: a view of an underlying range.

Reply:
This is not a view, but a structure containing a description of rows. Each element of the dm_rows vector contains a pointer to the start of the row data, its index and size. The segments() function needs to be reworked to make dm_rows work properly as a distributed array (in for_each etc.), and to get rid of custom functions.

Reply:
> The segments() function needs to be reworked

Is this described in some task and in some TODO in the code?